### PR TITLE
adapter: Support `COMMENT ON`

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR implements MVP support for the `COMMENT ON` function that is present in Postgres, behind a LaunchDarkly flag `enable_comment`. While theoretically we can leave comments for any kind of database object, this initial implementation only includes support for `TABLE`, `VIEW`, and `COLUMN`, I intend to add support for more types in a follow up PR.

It adds two new public facing features:

1. `COMMENT ON [ TABLE | VIEW | COLUMN ] <identfier> IS <comment>;`
2. BuiltinTable `mz_internal.mz_comments`

#### `mz_comments` schema

name        | type   | nullable
------------|--------|----------
id          | text   | false
object_type | text   | false
sub_id      | uint64 | true
comment     | text   | false

This is intended to mimic the [pg_description](https://www.postgresql.org/docs/current/catalog-pg-description.html) table where comments are stored in Postgres.

#### RBAC

[Postgres specifies](https://www.postgresql.org/docs/current/sql-comment.html) that only the owner of the object can leave a comment, [CockroachDB specifies](https://www.cockroachlabs.com/docs/stable/comment-on#required-privileges) that the user must have the `CREATE` privilege. I started by requiring both, and figured after discussion/dogfooding we can relax one or the other if need be.

#### Reading Comments

Postgres, Snowflake, and CoackroachDB all expose comments to users in a slightly different way.

Postgres: Meta commands and `pg_description` table
Snowflake: Returned as part of `SHOW ...` commands
CockroachDB: Returned as part of `SHOW ...` commands, if you specify `... WITH COMMENT`

I plan to add support for reading comments in a follow up PR after discussion.

### Motivation

* This PR adds a known-desirable feature.

  Closes https://github.com/MaterializeInc/materialize/issues/1212

### Tips for reviewer

This PR is broken up into parts that can be reviewed independently:

1. Updates to the SQL Parser to support the `COMMENT ON` syntax
4. New Stash Collection to store comments
5. Updates to planning for creating and updating comments
6. Updates to support deleting comments when dropping an object
7. Adding a LaunchDarkly flag, `enable_comment`
8. Testdrive and upgrade tests

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Adds the `COMMENT` feature present in Postgres behind a LaunchDarkly flag.

### CODE REVIEW

File: src/adapter/src/coord/command_handler.rs [hunk 1]
```diff
@@ -558,7 +558,9 @@ impl Coordinator {
                     | Statement::RevokePrivileges(_)
                     | Statement::RevokeRole(_)
                     | Statement::Update(_)
-                    | Statement::ValidateConnection(_) => {
+                    | Statement::ValidateConnection(_)
+                    // TODO(parkmycar): Reconsider COMMENT in transactions.
```
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:562]: I think this todo can go away: this should be the same as all other DDL.

File: src/adapter/src/catalog.rs [hunk 2]
```diff
@@ -6283,6 +6368,30 @@ impl Catalog {
                     );
                     builtin_table_updates.extend(state.pack_item_update(id, 1));
                 }
+                Op::Comment {
```
Comment (mjibson) [inline src/adapter/src/catalog.rs:6389]: Something needs to handle what happens to comments on drop/alter of the objects or columns. Are there dependencies we need to track? Can we just drop all comments on drop and not track dependencies?

**Review by mjibson — COMMENTED** [review]
Looking good.

Thoughts about reading comments: pg says "there is presently no security mechanism for viewing comments". We should not do that, but impose some read requirements on them.

**Comment by nrainer-materialize** [comment]
Since this PR touches quite some code and parser functionality, I would like to run a coverage run to detect possible test gaps. Please inform me once this PR is ready for that. Thank you!

File: src/sql-parser/src/ast/defs/statement.rs [hunk 3]
```diff
@@ -3701,3 +3704,80 @@ impl<T: AstInfo> AstDisplay for ReassignOwnedStatement<T> {
...
+        name: UnresolvedItemName,
+    },
+    Column {
+        relation_name: UnresolvedItemName,
+        column_name: Ident,
+    },
+}
+
+impl CommentObjectType {
+    pub fn name(&self) -> &UnresolvedItemName {
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql-parser/src/ast/defs/statement.rs]: It was only being called in one part of planning, I inlined the logic where it was called

File: src/adapter/src/catalog/storage.rs [hunk 4]
```diff
@@ -1775,6 +1795,43 @@ impl<'a> Transaction<'a> {
...
+        column_pos: Option<usize>,
+        comment: Option<String>,
+    ) -> Result<(), Error> {
+        let key = CommentKey {
+            object_id,
+            column_pos,
+        };
+        let exists = self.comments.get(&key).is_some();
+
+        match (exists, comment) {
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage.rs]: Good call! Updated to use `.set`!

File: src/adapter/src/catalog/storage.rs [hunk 5]
```diff
@@ -2103,6 +2162,17 @@ pub struct ItemValue {
     privileges: Vec<MzAclItem>,
 }
 
+#[derive(Clone, Debug, PartialOrd, PartialEq, Eq, Ord)]
+pub struct CommentKey {
+    object_id: CommentObjectId,
+    column_pos: Option<usize>,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage.rs]: Right, `column_pos` == `sub_component`. I renamed all of the sites to use `sub_component`

**Review by mjibson — APPROVED** [review]

**Comment by ParkMyCar** [comment]
Hey @nrainer-materialize you should be good to run a coverage test now!

**Comment by nrainer-materialize** [comment]
> Hey @nrainer-materialize you should be good to run a coverage test now!

Thank you for the note!
I triggered https://buildkite.com/materialize/coverage/builds/197 (but am aware that coverage builds are currently failing quite often and can't always be fixed by partial retries because that won't update coverage information). Let's see!

**Comment by ParkMyCar** [comment]
> Thank you for the note! I triggered https://buildkite.com/materialize/coverage/builds/197 

If I'm reading that right it looks like all changed lines are covered, and the changed lines only covered by a unit test are ones in the sql parser which is the right framework for such things. So I think we're good to go?

**Comment by nrainer-materialize** [comment]
> > Thank you for the note! I triggered https://buildkite.com/materialize/coverage/builds/197
> 
> If I'm reading that right it looks like all changed lines are covered, and the changed lines only covered by a unit test are ones in the sql parser which is the right framework for such things. So I think we're good to go?

Yes, it looks good. 👍 

If you want you could add end-to-end tests for ` impl AstDisplay for CommentStatement {fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {` and ` impl AstDisplay for CommentObjectType {
!    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {` but that's up to you if you consider that relevant or not.

**Pull request merged**